pageserver/src/tenant.rs

```diff
@@ -1421,24 +1509,85 @@ impl Tenant {
                 }
             }
 
-            // Ensure that there are no non-archived child timelines
-            let children: Vec<TimelineId> = timelines
-                .iter()
-                .filter_map(|(id, entry)| {
-                    if entry.get_ancestor_timeline_id() != Some(timeline_id) {
-                        return None;
-                    }
-                    if entry.is_archived() == Some(true) {
-                        return None;
-                    }
-                    Some(*id)
-                })
-                .collect();
+            if state == TimelineArchivalState::Archived {
+                // Ensure that there are no non-archived child timelines
+                let children: Vec<TimelineId> = timelines
+                    .iter()
+                    .filter_map(|(id, entry)| {
+                        if entry.get_ancestor_timeline_id() != Some(timeline_id) {
+                            return None;
+                        }
+                        if entry.is_archived() == Some(true) {
+                            return None;
+                        }
+                        Some(*id)
+                    })
+                    .collect();
 
-            if !children.is_empty() && state == TimelineArchivalState::Archived {
-                return Err(TimelineArchivalError::HasUnarchivedChildren(children));
+                if !children.is_empty() {
+                    return Err(TimelineArchivalError::HasUnarchivedChildren(children));
+                }
+            }
+            Some(Arc::clone(timeline))
+        };
+
+        let timeline = if let Some(timeline) = timeline_or_unarchive_offloaded {
+            timeline
+        } else {
+            let cancel = self.cancel.clone();
+            let timeline_preload = self
+                .load_timeline_metadata(timeline_id, self.remote_storage.clone(), cancel)
+                .await;
+
+            let index_part = match timeline_preload.index_part {
+                Ok(index_part) => {
+                    debug!("remote index part exists for timeline {timeline_id}");
+                    index_part
+                }
+                Err(DownloadError::NotFound) => {
+                    info!(%timeline_id, "index_part not found on remote");
```

Comment:
If the timeline is present in-memory but not remote storage, shouldn't that be unreachable?  (i.e. an error-level log message and a debug_assert)?

Reply:
> an earlier pageserver process can possibly delete it in the remote but not completely or something (say an older generation writing a file just after the runs the list operation).

Ah, yes.  This is a general problem with timeline deletion, where we rely on the property that any generation can safely delete a timeline because our caller (control plane) will never try to use a timeline after it is deleted.

But you're right, from the pageserver's point of view this is a possible case  & therefore we should warn but not assert.